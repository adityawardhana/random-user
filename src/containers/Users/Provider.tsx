import React, { FC, useEffect, useState } from "react";
import qs from "query-string";
import { getUsers } from "../../services/users";
import useFetch from "../../hooks/useFetch";
import useDebounce from "../../hooks/useDebounce";
import UserContext from "./Context";

const Provider: FC<{}> = ({ children }) => {
  const [params, setParams] = useState({
    inc: "gender,name,login,email,registered",
    results: 10,
    page: 1,
    keyword: "",
    sortBy: "",
    sortOrder: "",
    gender: "",
  });
  const [keyword, setKeyword] = useState("");
  const [searchFocus, setSearchFocus] = useState(false);

  const {
    fetch,
    loading,
    error,
    data: { results: users } = {},
  } = useFetch(
    (newParams) =>
      getUsers(qs.stringify(newParams || params, { skipEmptyString: true })),
    { immediate: false }
  );

  const handleChangeSearch = (e: React.ChangeEvent<HTMLInputElement>) => {
    setKeyword(e.target.value);
  };

  const handleSelectGender = (value: string) => {
    setParams({ ...params, gender: value });
  };

  const handleChangeSort = (sortBy: string, sortOrder: string) => {
    setParams({ ...params, sortBy, sortOrder, page: 1 });
  };

  const handleChangePage = (page: number) => {
    setParams({ ...params, page });
  };

  const handleResetFilter = () => {
    setKeyword("");
    setParams({ ...params, page: 1, keyword: "", gender: "" });
  };

  useDebounce(
    () => {
      searchFocus && setParams({ ...params, keyword, page: 1 });
    },
    keyword,
    1000
  );

  useEffect(() => {
    fetch({ ...params });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [params]);

  return (
    <UserContext.Provider
      value={{
        params,
        users,
        loading,
        error,
        keyword,
        setSearchFocus,
        handleChangePage,
        handleChangeSearch,
        handleChangeSort,
        handleSelectGender,
        handleResetFilter,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

export default Provider;
