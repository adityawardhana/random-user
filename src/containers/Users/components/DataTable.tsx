import { Fragment } from "react";
import { format } from "date-fns";
import Sorter from "../../../ui/Sorter";
import Table from "../../../ui/Table";
import useUsers from "../useUser";

const DataTable = () => {
  const {params, handleChangeSort, loading, users} = useUsers()
  return (
    <Table shadow round responsive>
      <thead>
        <tr>
          <th>
            <Sorter
              name="Username"
              value="username"
              order={params.sortOrder}
              onChange={handleChangeSort}
              selectedValue={params.sortBy}
            />
          </th>
          <th>
            <Sorter
              name="Name"
              value="name"
              order={params.sortOrder}
              onChange={handleChangeSort}
              selectedValue={params.sortBy}
            />
          </th>
          <th>
            <Sorter
              name="Email"
              value="email"
              order={params.sortOrder}
              onChange={handleChangeSort}
              selectedValue={params.sortBy}
            />
          </th>
          <th>
            <Sorter
              name="Gender"
              value="gender"
              order={params.sortOrder}
              onChange={handleChangeSort}
              selectedValue={params.sortBy}
            />
          </th>
          <th>
            <Sorter
              name="Registered Date"
              value="registered.date"
              order={params.sortOrder}
              onChange={handleChangeSort}
              selectedValue={params.sortBy}
            />
          </th>
        </tr>
      </thead>
      {!loading && (
        <tbody>
          {users?.map((user: any, index: number) => {
            return (
              <Fragment key={index}>
                <tr>
                  <td>{user.login?.username}</td>
                  <td>
                    {user.name?.first} {user.name?.last}
                  </td>
                  <td>{user.email}</td>
                  <td>{user.gender}</td>
                  <td>
                    {format(
                      new Date(user.registered?.date),
                      "dd-MM-yyyy HH:mm"
                    )}
                  </td>
                </tr>
              </Fragment>
            );
          })}
        </tbody>
      )}
    </Table>
  );
};

export default DataTable;
