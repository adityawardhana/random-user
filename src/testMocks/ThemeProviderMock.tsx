import React from "react";
import theme, { GlobalStyles } from "@adityawardhana/ts-react-ui/lib/theme";
import { ThemeProvider } from "styled-components";

export const setupTheme = (children: React.ReactNode) => {
  return (
    <ThemeProvider theme={theme}>
      <GlobalStyles />
      {children}
    </ThemeProvider>
  );
};
