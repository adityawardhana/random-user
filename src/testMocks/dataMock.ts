export const resultAPIMock = {
  login: {
    username: "adityawrdhn",
  },
  name: {
    first: "adit",
    last: "w",
  },
  email: "adityawrdhn@gmail.com",
  gender: "male",
  registered: { date: "2003-08-26T21:56:32.575Z" },
};

export const usersMock = {
  params: { sortBy: "email", sortOrder: "ascend", page: 1, gender: "" },
  loading: false,
  users: [...Array(3)].map(() => ({ ...resultAPIMock })),
  keyword: "xxx",
  handleChangeSort: jest.fn(),
  handleChangePage: jest.fn(),
  handleChangeSearch: jest.fn(),
  setSearchFocus: jest.fn(),
  handleSelectGender: jest.fn(),
  handleResetFilter: jest.fn(),
};
