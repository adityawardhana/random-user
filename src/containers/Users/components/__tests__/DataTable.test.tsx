import { render } from "@testing-library/react";
import { usersMock } from "../../../../testMocks/dataMock";
import { setupTheme } from "../../../../testMocks/ThemeProviderMock";
import Provider from "../../Provider";
import useUsersMock from "../../useUser";
import DataTable from "../DataTable";

jest.mock("../../useUser", () => ({
  __esModule: true,
  default: jest.fn(),
}));

describe("Data Table unit tests", () => {
  beforeEach(() => {
    (useUsersMock as any).mockReturnValue({ ...usersMock });
  });
  test("should render correctly", () => {
    const { container } = render(
      setupTheme(
        <Provider>
          <DataTable />
        </Provider>
      )
    );
    expect(container).toMatchSnapshot();
  });
});
