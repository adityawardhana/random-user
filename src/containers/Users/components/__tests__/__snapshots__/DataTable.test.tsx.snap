// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data Table unit tests should render correctly 1`] = `
.c2 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c3 {
  font-size: 12px;
  font-weight: 400;
  line-height: 16px;
  color: #152E3B;
  white-space: normal;
  font-weight: bold;
}

.c5 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 16px;
  height: 16px;
  pointer-events: none;
}

.c5[type="button"] .c4 {
  cursor: pointer;
}

.c5 svg {
  width: 16px;
  height: 16px;
  fill: #152E3B;
}

.c0 {
  display: block;
  width: 100%;
  overflow-x: auto;
  border-radius: 4px;
  -moz-border-radius: 4px;
  -webkit-border-radius: 4px;
}

.c0.shadow {
  box-shadow: 0 1px 4px 0 rgba(112,114,125,0.4);
}

.c0 table {
  width: 100%;
}

.c0::-webkit-scrollbar {
  height: 8px;
}

.c0::-webkit-scrollbar-thumb {
  background-color: #dddddd;
  border-radius: 10px;
}

.c0::-webkit-scrollbar-track {
  background-color: #e6e6e659;
}

.c0 {
  box-shadow: 0 1px 4px 0 rgba(112,114,125,0.4);
}

.c1 {
  background-color: #ffffff;
  border-radius: 4px;
  -moz-border-radius: 4px;
  -webkit-border-radius: 4px;
  border-collapse: separate;
}

.c1 thead th {
  background-color: #ffffff;
  color: #70727d;
}

.c1 tbody tr {
  background-color: #f8fbfc;
}

.c1 tbody tr:nth-child(2n) {
  background-color: #ffffff;
}

.c1 th:first-of-type {
  border-radius: 4px 0 0 0;
  -moz-border-radius: 4px 0 0 0;
  -webkit-border-radius: 4px 0 0 0;
}

.c1 th:last-of-type {
  border-radius: 0 4px 0 0;
  -moz-border-radius: 0 4px 0 0;
  -webkit-border-radius: 0 4px 0 0;
}

.c1 tr:last-of-type td:first-of-type {
  border-radius: 0 0 0 4px;
  -moz-border-radius: 0 0 0 4px;
  -webkit-border-radius: 0 0 0 4px;
}

.c1 tr:last-of-type td:last-of-type {
  border-radius: 0 0 4px 0;
  -moz-border-radius: 0 0 4px 0;
  -webkit-border-radius: 0 0 4px 0;
}

.c1 td:first-of-type,
.c1 th:first-of-type {
  border-left: medium none;
}

.c1 th,
.c1 td {
  padding: 14px 24px;
  font-size: 12px;
  vertical-align: middle;
}

.c1 th[rowspan],
.c1 td[rowspan] {
  vertical-align: middle;
}

<div>
  <div
    class="c0"
  >
    <table
      class="c1"
    >
      <thead>
        <tr>
          <th>
            <span
              class="c2"
              cursor="pointer"
            >
              <span
                class="c3"
                color="black900"
                type="B12"
              >
                Username
              </span>
            </span>
          </th>
          <th>
            <span
              class="c2"
              cursor="pointer"
            >
              <span
                class="c3"
                color="black900"
                type="B12"
              >
                Name
              </span>
            </span>
          </th>
          <th>
            <span
              class="c2"
              cursor="pointer"
            >
              <span
                class="c3"
                color="black900"
                type="B12"
              >
                Email
              </span>
              <i
                class="c4 c5"
                color="secondary"
              >
                <svg
                  aria-hidden="true"
                  class="octicon octicon-chevron-up"
                  fill="currentColor"
                  height="16"
                  role="img"
                  style="display: inline-block; user-select: none; vertical-align: text-bottom;"
                  viewBox="0 0 16 16"
                  width="16"
                >
                  <path
                    d="M3.22 9.78a.75.75 0 010-1.06l4.25-4.25a.75.75 0 011.06 0l4.25 4.25a.75.75 0 01-1.06 1.06L8 6.06 4.28 9.78a.75.75 0 01-1.06 0z"
                    fill-rule="evenodd"
                  />
                </svg>
              </i>
            </span>
          </th>
          <th>
            <span
              class="c2"
              cursor="pointer"
            >
              <span
                class="c3"
                color="black900"
                type="B12"
              >
                Gender
              </span>
            </span>
          </th>
          <th>
            <span
              class="c2"
              cursor="pointer"
            >
              <span
                class="c3"
                color="black900"
                type="B12"
              >
                Registered Date
              </span>
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            adityawrdhn
          </td>
          <td>
            adit
             
            w
          </td>
          <td>
            adityawrdhn@gmail.com
          </td>
          <td>
            male
          </td>
          <td>
            27-08-2003 04:56
          </td>
        </tr>
        <tr>
          <td>
            adityawrdhn
          </td>
          <td>
            adit
             
            w
          </td>
          <td>
            adityawrdhn@gmail.com
          </td>
          <td>
            male
          </td>
          <td>
            27-08-2003 04:56
          </td>
        </tr>
        <tr>
          <td>
            adityawrdhn
          </td>
          <td>
            adit
             
            w
          </td>
          <td>
            adityawrdhn@gmail.com
          </td>
          <td>
            male
          </td>
          <td>
            27-08-2003 04:56
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
